metricbeat/module/elasticsearch/index/data_xpack.go

```diff
@@ -0,0 +1,332 @@
+
+	// "index_stats.created",
+	v, err := indexMetadata.GetValue("settings.index.creation_date")
+	if err != nil {
+		return err
+	}
+	c, ok := v.(string)
+	if !ok {
+		return elastic.MakeErrorForMissingField("settings.index.creation_date", elastic.Elasticsearch)
+	}
+
+	created, err := strconv.ParseInt(c, 10, 64)
+	if err != nil {
+		return err
+	}
+	indexStats.Put("created", created)
+
+	// "index_stats.version.created", <--- don't think this is being used in the UI, so can we skip it?
+	// "index_stats.version.upgraded", <--- don't think this is being used in the UI, so can we skip it?
+
+	// "index_stats.status",
+	s, err := indexRoutingTable.GetValue("shards")
+	if err != nil {
+		return err
+	}
+	shards, ok := s.(map[string]interface{})
+	if !ok {
+		return elastic.MakeErrorForMissingField("shards", elastic.Elasticsearch)
+	}
+
+	status, err := getIndexStatus(shards)
+	if err != nil {
+		return err
+	}
+	indexStats.Put("status", status)
+
+	// "index_stats.shards.total",
+	// "index_stats.shards.primaries",
+	// "index_stats.shards.replicas",
+	// "index_stats.shards.active_total",
+	// "index_stats.shards.active_primaries",
+	// "index_stats.shards.active_replicas",
+	// "index_stats.shards.unassigned_total",
+	// "index_stats.shards.unassigned_primaries",
+	// "index_stats.shards.unassigned_replicas",
+	// "index_stats.shards.initializing",
+	// "index_stats.shards.relocating",
+	shardStats, err := getIndexShardStats(shards)
+	if err != nil {
+		return err
+	}
+	indexStats.Put("shards", shardStats)
+	return nil
+}
+
+func getClusterStateMetricForIndex(clusterState common.MapStr, index, metricKey string) (common.MapStr, error) {
+	fieldKey := metricKey + ".indices." + index
+	value, err := clusterState.GetValue(fieldKey)
+	if err != nil {
+		return nil, err
+	}
+
+	metric, ok := value.(map[string]interface{})
+	if !ok {
+		return nil, elastic.MakeErrorForMissingField(fieldKey, elastic.Elasticsearch)
+	}
+	return common.MapStr(metric), nil
+}
+
+func getIndexStatus(shards map[string]interface{}) (string, error) {
+	if len(shards) == 0 {
+		// No shards, index is red
+		return "red", nil
+	}
+
+	areAllPrimariesStarted := true
+	areAllReplicasStarted := true
+
+	for _, indexShard := range shards {
+		is, ok := indexShard.([]interface{})
+		if !ok {
+			return "", fmt.Errorf("shards is not an array")
+		}
+
+		for _, shard := range is {
+			s, ok := shard.(map[string]interface{})
+			if !ok {
+				return "", fmt.Errorf("shards is not an array of shard objects")
+			}
+
+			shard := common.MapStr(s)
+
+			isPrimary := shard["primary"].(bool)
+			state := shard["state"].(string)
+
+			if isPrimary {
+				areAllPrimariesStarted = areAllPrimariesStarted && (state == "STARTED")
+			} else {
+				areAllReplicasStarted = areAllReplicasStarted && (state == "STARTED")
+			}
+		}
+	}
+
+	if areAllPrimariesStarted && areAllReplicasStarted {
+		return "green", nil
+	}
+
+	if areAllPrimariesStarted && !areAllReplicasStarted {
+		return "yellow", nil
+	}
```

Comment:
Could we get the index status directly from ES? I feel like we are re-implementing here the logic to decide the status of an index, and this forces us to keep this in sync with ES implementation (even if it is simple and I don't think this is going to change frequently :slightly_smiling_face:)

Reply:
Yeah, I agree this would be ideal to get from ES itself. I made this implementation here as a fallback because we'll need to get the ES team to modify the cluster state API response, which they may or may not like to do given how core it is. Anyway, it's worth asking at least so I will do that.

Reply:
I put up a PR for the Elastiscearch Cluster State API to return index status in it: https://github.com/elastic/elasticsearch/pull/33614. Let's see what the ES team says.